app/javascript/extensible-components/index.ts

```diff
@@ -0,0 +1,69 @@
+import { IExtensionComponent, IMiQApiCallback } from './lib';
+
+const source = new Rx.Subject();
+const items = new Set();
+
+/**
+ * Class for easy creation of extensible component.
+ */
+export class ExtensibleComponent {
+  public unsubscribe: () => void;
+  constructor(public name: string, public api: IMiQApiCallback, public render: IMiQApiCallback){}
+}
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
+ */
+function addComponent(name: string, api?: IMiQApiCallback, render?: IMiQApiCallback) {
+  let newCmp = new ExtensibleComponent(name, api, render);
+  source.onNext({action: 'add', payload: newCmp});
+  return newCmp;
+}
+
+/**
+ * Helper function to subscribe to extension items based on component's name.
+ * @param cmpName name of component we want to subscribe to.
+ * @return object which has with and unsubscribe property. With is for callback to use found component and delete to
+ * unsubscribe from rxjs subject.
+ */
+function subscribe(cmpName: string) {
+  let unsubscribe;
+  return {
+    with: (callback) => {
+      unsubscribe = source
+        .map(sourceAction => sourceAction.action === 'add' ? sourceAction.payload : {})
+        .filter(component => component && component.name === cmpName)
+        .subscribe(cmp => cmp && callback(cmp));
+
+      ManageIQ.extensions.items.forEach((component) => {
+        component.name === cmpName && callback(component);
+      });
+    },
+    delete: () => unsubscribe && unsubscribe()
+  }
+}
+
+const extensions = {
+  addComponent: addComponent,
+  subscribe: subscribe,
+  getItems: () => items
+}
+
+ManageIQ.extensions = ManageIQ.extensions || extensions;
+
+/**
+ * Subscribe to extensions source to add new components to items object.
+ */
+source.subscribe((component: IExtensionComponent) => {
+  if (component.action === 'add' && component.hasOwnProperty('payload')) {
+    component.payload.unsubscribe = () => {
```

Comment:
Regarding the `ExtensibleComponent`'s `unsubscribe` property - this is part of the component's API, just like `api` and `render` objects.

I'd suggest to rename it to `delete` or `remove` because it's used to remove the component from `items` collection.

Reply:
I also made it shorter, so the delete method will return true if component was removed from set.